### PR TITLE
Add support of snapshot in vSphere virtualmachine metricset

### PR DESCRIPTION
**Pull request by ishleenk17**
Add support of adding snapshot details to the Virtualmachine object in Vsphere.
It caters to the snapshot and any childsnapshots as well. 

Relates: https://github.com/elastic/obs-integration-team/issues/34

### CODE REVIEW

**Comment by niraj-elastic** [comment]
@ishleenk17 please add entry in ascii.next file.

**Review by niraj-elastic — CHANGES_REQUESTED** [review]

**Comment by niraj-elastic** [comment]
Please update `metricbeat/module/vsphere/virtualmachine/_meta/data.json` as well.

**Comment by ishleenk17** [comment]
> Please update `metricbeat/module/vsphere/virtualmachine/_meta/data.json` as well.

There is no way to test this, due to unavailability of snapshots. Hence its not updated. 
An option is to put the dummy data in the sample.

File: metricbeat/module/vsphere/virtualmachine/_meta/fields.yml [hunk 1]
```diff
@@ -91,5 +91,11 @@
       type: long
       description: >
         The uptime of the VM in seconds.
+    - name: snapshots
```
[the remark this answers is not in the document]
Comment (ishleenk17) [reply metricbeat/module/vsphere/virtualmachine/_meta/fields.yml:94]: Hmmm, yes we can. Let me do that.

**Comment by niraj-elastic** [comment]
> > Please update `metricbeat/module/vsphere/virtualmachine/_meta/data.json` as well.
> 
> There is no way to test this, due to unavailability of snapshots. Hence its not updated. An option is to put the dummy data in the sample.

I understand, But the main prepose of data.json is to give idea about event that is going to be produced. I think we should add dummy fields because it would capture structure of our event.

**Comment by shmsr** [comment]
> > > Please update `metricbeat/module/vsphere/virtualmachine/_meta/data.json` as well.
> > 
> > 
> > There is no way to test this, due to unavailability of snapshots. Hence its not updated. An option is to put the dummy data in the sample.
> 
> I understand, But the main prepose of data.json is to give idea about event that is going to be produced. I think we should add dummy fields because it would capture structure of our event.

I think it is possible to write a mock (see [example](https://github.com/vmware/govmomi/blob/5d39eba0ed5eb3cdbc39f0f85d610684168a17db/object/example_test.go#L65)) but given the strict timeline we can take it up in future to properly generate the data.json. For now, we can go with dummy.

File: metricbeat/module/vsphere/virtualmachine/data_test.go [hunk 2]
```diff
@@ -58,6 +58,16 @@ func TestEventMapping(t *testing.T) {
...
 			"customField2": "value2",
 		},
+		Snapshots: []VMSnapshotData{
+			{
+				Name:        "Snapshot_1",
+				Description: "Test snapshot 1",
+			},
+			{
+				Name:        "Snapshot_2",
+				Description: "Test snapshot 2",
```
[the remark this answers is not in the document]
Comment (ishleenk17) [reply metricbeat/module/vsphere/virtualmachine/data_test.go:70]: Added

**Comment by shmsr** [comment]
```diff
diff --git a/metricbeat/module/vsphere/virtualmachine/data.go b/metricbeat/module/vsphere/virtualmachine/data.go
index 69d0bd1df5..bc29c1f8ef 100644
--- a/metricbeat/module/vsphere/virtualmachine/data.go
+++ b/metricbeat/module/vsphere/virtualmachine/data.go
@@ -72,9 +72,9 @@ func (m *MetricSet) mapEvent(data VMData) mapstr.M {
 	if len(data.DatastoreNames) > 0 {
 		event["datastore.names"] = data.DatastoreNames
 	}
-	if len(data.Snapshots) > 0 {
+	if n := len(data.Snapshots); n > 0 {
 		event["snapshot.info"] = data.Snapshots
-		event["snapshot.count"] = len(data.Snapshots)
+		event["snapshot.count"] = n
 	}
 
 	return event
diff --git a/metricbeat/module/vsphere/virtualmachine/virtualmachine.go b/metricbeat/module/vsphere/virtualmachine/virtualmachine.go
index 81052e1eed..071c4f6cfb 100644
--- a/metricbeat/module/vsphere/virtualmachine/virtualmachine.go
+++ b/metricbeat/module/vsphere/virtualmachine/virtualmachine.go
@@ -61,6 +61,10 @@ type VMData struct {
 }
 
 type VMSnapshotData struct {
+	ID          int32
+	State       string
+	Quiesced    bool
+	Age         time.Duration
 	Name        string
 	Description string
 	CreateTime  time.Time
@@ -286,9 +290,12 @@ func getHostSystem(ctx context.Context, c *vim25.Client, ref types.ManagedObject
 }
 
 func fetchSnapshots(snapshotTree []types.VirtualMachineSnapshotTree) []VMSnapshotData {
-	var snapshots []VMSnapshotData
+	snapshots := make([]VMSnapshotData, 0, len(snapshotTree))
 	for _, snapshot := range snapshotTree {
 		snapshots = append(snapshots, VMSnapshotData{
+			ID:          snapshot.Id,
+			State:       string(snapshot.State),
+			Quiesced:    snapshot.Quiesced,
 			Name:        snapshot.Name,
 			Description: snapshot.Description,
 			CreateTime:  snapshot.CreateTime,
```

Code added looks good. Just have a couple of things to suggest as you can see from the patch:

- IMO, adding ID and State makes sense for Snapshot as ID is the unique identifier for the Snapshot and State gives the info if the snapshot is powered on/off or suspended. Quiesced seems also important but not sure how much:

Quiesced definition:
```
// Flag to indicate whether or not the snapshot was created with
// the "quiesce" option, ensuring a consistent state of the file system.
```

- Also, I did some safe preallocation for the slice; fewer allocations now than before.

**Review by shmsr — APPROVED** [review]
Left one comment to address. Everything else looks good. So, approving!

**Comment by ishleenk17** [comment]
> * IMO, adding ID and State makes sense for Snapshot as ID is the unique identifier for the Snapshot and State gives the info if the snapshot is powered on/off or suspended. Quiesced seems also important but not sure how much:

State and ID could be helpful, so adding those. Quiesced is not required IMO.

File: CHANGELOG.next.asciidoc [hunk 3]
```diff
@@ -60,6 +60,7 @@ https://github.com/elastic/beats/compare/v8.8.1\...main[Check the HEAD diff]
 - Update metrics for the vSphere Host metricset. {pull}40429[40429]
 - Mark system process metricsets as running if metrics are partially available {pull}40565[40565]
 - Added back `elasticsearch.node.stats.jvm.mem.pools.*` to the `node_stats` metricset {pull}40571[40571]
+- Add support for snapshot in virtualmachine {pull}40683[40683]
```
Comment (shmsr) [inline CHANGELOG.next.asciidoc:63]: nitpick: As this CHANGELOG gets published; let's also vSphere.

maybe:
> vSphere: Add support for snapshot in virtualmachine

or:

> metricbeat/module/vsphere/virtualmachine: Add support for snapshot

Same change in PR title as well

**Review by niraj-elastic — APPROVED** [review]
LGTM

**Comment by shmsr** [comment]
> @ishleenk17 I was working on alerts and i found that the way in which we define snapshot field, does not work. It will create field with name `snapshot.info` also since name in objects are defined with capital letters the dynamic fields will also be capital. For example `"snapshot.info": { "ID": "xxxx" }`
> 
> We have to change it.

I agree with this point to make it small-case. But I'm curious; if field names are not in small-case, alerts don't work?

**Pull request merged**